### PR TITLE
expression,util/types: update time Add() implementation

### PR DESCRIPTION
**Pull request by tiancaiamao**
Time add operation should not convert the date to Go time, calculate and convert the date back.
It's a dirty hack, ugly and error prone, so I update the implementation.
We won't bother by timezone involved in conversion any more.
Fix https://github.com/pingcap/tidb/issues/3944

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -1946,17 +1945,13 @@ func strDatetimeAddDuration(d string, arg1 types.Duration) (result types.Datum,
 	if err != nil {
 		return result, errors.Trace(err)
 	}
-	tmpDuration := arg0.Add(arg1)
-	resultDuration, err := tmpDuration.ConvertToTime(mysql.TypeDatetime)
-	if err != nil {
-		return result, errors.Trace(err)
-	}
+	tmp := arg0.Add(arg1)
```
Comment (shenli) [inline expression/builtin_time.go]: Add check()

File: util/types/time.go [hunk 2]
```diff
@@ -486,12 +486,22 @@ func (t *Time) Sub(t1 *Time) Duration {
...
+	sign, hh, mm, ss, micro := splitDuration(d.Duration)
+	seconds, microseconds, _ := calcTimeDiff(t.Time, FromDate(0, 0, 0, hh, mm, ss, micro), -sign)
+	days := seconds / secondsIn24Hour
+	year, month, day := getDateFromDaynr(uint(days))
+	var tm mysqlTime
+	tm.year, tm.month, tm.day = uint16(year), uint8(month), uint8(day)
+	calcTimeFromSec(&tm, seconds%secondsIn24Hour, microseconds)
+	ret := Time{
+		Time:     tm,
+		Type:     t.Type,
```
Comment (XuHuaiyu) [inline util/types/time.go]: If t.Type == mysql.TypeDate
we still set the type of result to mysql.TypeDate?

Comment (tiancaiamao) [reply util/types/time.go]: Yes.
Do you mean hh:mm:ss should be truncated?

**Comment by zimulala** [comment]
Please fix this conflict. @tiancaiamao

**Comment by tiancaiamao** [comment]
Done @zimulala

**Comment by zimulala** [comment]
LGTM

**Pull request merged**